### ISSUE TITLE
No description in VS2019

### ISSUE THREAD

**Issue by Evangelink** [deviousasti/fsharp-formatting-for-vs#9]
Hello,

I am new to using this extension and I have just noticed that there is no description for the entries under `Visual Studio 2019` options:

![image](https://user-images.githubusercontent.com/11340282/101625554-dbbc2280-3a1b-11eb-95e6-58831961f663.png)

**Comment by deviousasti** [comment]
I've been meaning to put that in for some time, although it's not very useful without concrete examples for before and after.
The options are here:
https://github.com/fsprojects/fantomas/blob/master/docs/Documentation.md

Do you have any good suggestions?

**Comment by Evangelink** [comment]
Ideally, having the description and a dynamic example which changes based on the value of the setting as it is done for ReSharper or C# code styles.

![image](https://user-images.githubusercontent.com/11340282/101633905-3a879900-3a28-11eb-8de0-688b7f8c9ad6.png)

![image](https://user-images.githubusercontent.com/11340282/101633920-43786a80-3a28-11eb-917d-cf74e6b3ecf8.png)

I would be happy to help here, this way I could learn the options and see if I want some tweaks to the default configuration :)

**Comment by deviousasti** [comment]
Any sort of help is appreciated!
These are the default options to be used if there's no `.editorconfig` file.

This is where the descriptions are specified.
https://github.com/deviousasti/fsharp-formatting-for-vs/blob/bdb75fa49d01a3eb5b6f6a84ad5dcb71a634ca16/src/FantomasOptionsPage.cs#L204

**Comment by Evangelink** [comment]
Let me dive back into VS option pages (I haven't used them in a while).

**Comment by deviousasti** [comment]
If you're free this weekend, we can work together on this. VS live share session?

**Comment by Evangelink** [comment]
Sorry @deviousasti I missed this notif, I might have some availability next week-end or at evening (CET timezone).

**Comment by Evangelink** [comment]
Regarding the `Description`, most of the entries are missing the `[Description(XXX)]` attribute which explains why there is no example.

I am still trying to find how to have the dynamic code example, looking at roslyn and some other repositories we might have to use a WPF page instead of WinForms.

**Comment by deviousasti** [comment]
Oh, sorry, I forgot to explain.
We're using this attribute to specify that VS needs to use this as an options page:

https://github.com/deviousasti/fsharp-formatting-for-vs/blob/bdb75fa49d01a3eb5b6f6a84ad5dcb71a634ca16/src/FantomasVsPackage.cs#L27
Which shows up the options in a property editor.

Adding a description attribute is one step forward, but it isn't sufficient for the Roslyn like experience. 
We're going to have to recreate the configuration UI (checkboxes, dropdown, etc) in a WPF UserControl, and then add a set of examples for the options.

I've been working with WPF for 10 years now, so not a big deal.
If we add a custom attribute for the corresponding ` .editorconfig` option, we can have a nice UI to generate an ` .editorconfig` file as well.

**Comment by Evangelink** [comment]
Yeah the only problem for me so far is just to be able to include a `TextEditor` like component which will reuse all the formattings and colors from the F# configuration in the IDE. Maybe I am going a little too far and we could simply start with just a simple text (no colors + no IDE specific background color).

**Comment by deviousasti** [comment]
Yeah, that's probably a bit extreme for now. I have seen other extensions use [AvalonEdit](https://github.com/icsharpcode/AvalonEdit) for displaying code, but that's a lot of stuff for this one feature.

If the examples are simple enough, coloring needn't be an issue.
Although... we're dependent on FCS and have to parse the examples anyway, so we could probably assign a colour per token, and it would be reasonable.

**Comment by nojaf** [comment]
Hey Asti, I'm afraid I can't really add anything useful here.

Rider uses a custom editorconfig parsing thing, they added attributes for each setting as well: https://github.com/JetBrains/fsharp-support/blob/93ab17493a34a0bc0fd4c70b11adde02f81455c4/ReSharper.FSharp/src/FSharp.Psi/src/CodeFormatter/FSharpFormatSettingsKey.cs

Ionide does not have any UI for the settings.

Fantomas Online has a `copy settings` button when you change the defaults from the menu.
[Example](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKAbRAXMBDFAlgHY4C8YADJADThQDO2WxA5vamANp0SgT9QAJFgCeAB2RpIxLDR79IggG650AV0ld5AirQEDIASSJISAZQIAvZHv0QALNrABdeQF9bvJ4tESO0kjk7RRV1TW47CABGT31IAFlcAA8AGWJEFMQiFiwACyDIqIAmCidXAQ95PjjhcU1IACNYWHQCg2VVDQ4IyKLYgzNEAFsCOHRYIgBBLABREwB5ADM0ohsnCEXVRjL3WOr233qmlraFDrDu9bAAZn6FMzFcaEQAIURF2AAnRAAFXE-cENsIhPqcBFhPhodhU9t5an4pMdWncQp1wld7HcIJAHk9Xu8vhlYAB3EHQXCMYxKWDk5gTMH8CFQuzlfiVAT7M6HfxIhmoi5oHp2ACsWKguOebw%2B3wAqmIJJ9yZSiNTaQR6WLNuhtizdlU4dzEc1kXDQl1BVcAGxinGPSUE74AYXQFPojom9CZ0CwXwZGy2iGhbNhwXhR2NfPO5q0kTAAHYbRL8dLEPFhg0QX6wFqdfpWRB2fxOdiwzyIyio%2BjYwAORN25OE93jIhZpmB3Uw-Whw1QXkVs1VyIATjreMmiywIPdQyGuFbkPbeb1HINdTLJ37aMusaiuiutrxUsJgxGYw1Vxzi4E%2BbAha83bXRo3pq3Fp3MX3SfHk8%2BJ9GLXPWM2yDAsQxqHtGnLF8BRjQo%2Bn3YxTCweYiAAFU%2BEQAHUCDyLNLxA28wIOR9eyg0MB23Qpbk-etJk%2BWA1BMAARRB0AIEYf3nZkl07FcHwRKAZEjCi30KTF90SJJDEWVDciyGZtUQMxci%2BLBsPgXCxXsUoO2DLtwJIgJZE3GChX0KJRQk5JjEWAgknmBVcB9T4ZiSMRvnoeh1RbMVhR0ni9L4gyBKM4TX1guwomtKykgAJUQOBPngdTNIxfzr2XItVxCoSTOjMyBCiBMYvixL4AAOTUIYM0%2BJZDEnIZ2DFKICLvMBiyECChjUdBmDY1YADEvlnLAf1Qtc8sHSLa33UqvngeIer69Ihs%2BEbOJtaBcn%2BJ4fwAfWJAgNPyVqiK5Qzcug-KriiEcYrogERHmT40k9FL8i09L%2BBvNqOp8C7Akmyi7BKG1JIe3AnpeghPUq6qQTqhqmpu079OIkLut6gh%2BsQVb1pBca-CB0SQY-WNIAhqHXqwRasZxvGnI2-ctp270QQOo7cNRoL0fqS7yPCgr%2BCKeDyckgA1NEXmIeBWHerNtO5rL%2BL5wGrqm-QimosXkgGxjvW86WTDlzmPrSpX72C1XjPV4HNfEnWkiY2AsAAcWwVz3MQTzvPl3yvogH6zpLCC%2B1tknNcssWlux9IXnGaAAGsXgBJPsHoFCzEBRAmyqnyLwDC32uy8NnwF0yriKaLyfKxBiRxt4sFJLJCcQFjbKIHDvMmEw03hz5kdjfDdNAtHzpCsPy%2Bu2MimK8mAGlEEQMRpNkrJjCzoEVjWIfC5Hwix5DgGbanjWBCKGbHYUkZ6FyP3zf336S-XE1T7t8%2B7vJixskwOiWCqrIalEANHiAgHekRh4BVHjzcepdX41BEhFfQ1w9zk0mGxFgRA9ZEANhMCwmCnJqG%2BKhWAiFAFOW8nhPeUCD4wKPhPMiCDBZXGuGTSIFNeogiIBQpQRJsh9xqu3YgXcPRUMUkXP6pYnzwPaIgoWEBrii3YbTZg28UiAgaPAXAzpYBeWyLXeu6QxG5gyrxZWVsX5hQrrGa42t2FMRhrgBomBr4w1yGYEQJBkjGKvN9TKlteb%2BCyPAJY28zCiEwFY6ekRrgO3YXMEJywjGbU%2BOgRYkAJHP2kVEs%2B-BrhR3YeEz4owaZgJ8a1OgzhbDSHoANLyHB8JuCAA) 
That copies the `.editorconfig` settings to the clipboard.
But again no description.

**Comment by Evangelink** [comment]
@deviousasti How would you like to proceed, I start some PR and we contribute together?

Also I am wondering if you want to hardcode the various properties or if you want to actually read some configuration file and dynamically generate the settings page?

**Comment by deviousasti** [comment]
Well, no we can have some custom attribute on each property and use reflection to generate our UI.

> Yeah the only problem for me so far is just to be able to include a TextEditor like component
I made this so we can have our cake and eat it too.

https://github.com/deviousasti/fsharp-highlighting

**Comment by deviousasti** [comment]
As for the your branch, would I have commit access?

**Comment by Evangelink** [comment]
Sure!
